sympy/ntheory/residue_ntheory.py

```diff
@@ -615,16 +624,47 @@ def is_nthpow_residue(a, n, m):
 
     P. Hackman "Elementary Number Theory" (2009),  page 76
     """
-    if( primitive_root(m) == None ):
-        raise NotImplementedError("Not implemented for m which do not have any primitive root")
+    if n == 0:
+        if m == 1:
+            return False
+        return a == 1
     if n == 1:
         return True
     if n == 2:
         return is_quad_residue(a, m)
+    if a == m:
+        return bool(n)
+    if primitive_root(m) is None:
+        return _nthroot_mod_prime_power(a, n, m)
     f = totient(m)
     k = f // igcd(f, n)
     return pow(a, k, m) == 1
 
+
+def _nthroot_mod_prime_power(a, n, m):
+    """Returns true if solution for ``x**n == a (mod(p**k))`` exists."""
+    for p, k in factorint(m).items():
+        if a % p:
+            if p != 2:
+                return is_nthpow_residue(a, n, pow(p, k))
+            if k == 1:
+                return True
+            if n & 1:
+                return True
+            c = trailing(n)
+            if k == 2:
+                return a % 4 != 3
```

Comment:
I think this could be changed to ``return a % 4 == 1`` and then merged with the next line. (That could even take care of the case  k = 1.)